sympy/tensor/array/sp_utils.py

```diff
@@ -0,0 +1,396 @@
+from sympy.core.sympify import _sympify
+from sympy import Tuple, Basic, S
+from sympy.tensor.array.ndim_array import ImmutableNDimArray
+from sympy.tensor.array.sparse_ndim_array import SparseNDimArray
+from sympy.tensor.array.dense_ndim_array import DenseNDimArray
+from sympy.core.sympify import sympify
+from sympy.core.compatibility import Iterable
+
+import functools
+
+
+class SparseArrayFormat(SparseNDimArray, ImmutableNDimArray):
+    """
+    A base class for all sparse array formats. It is supposed to be Immutable.
+    """
+    def _check_bound(self, shape, index):
+        if len(shape) != len(index):
+            return False
+        return all([shape[i] > index[i] for i in range(len(shape))])
+
+    # A test for mul and rmul, this code should be added to the original
+    # function once the test is valide
+    def __mul__(self, other):
+        other = sympify(other)
+        return self.toscr()._mul(other)
+
+    def __setitem__(self, index, value):
+        raise TypeError('Immutable N-dim array')
```

Comment:
what about moving this to the `Immutable...` superclass?

Reply:
Yes, this line of code is also used in `ImmutableDenseNDimArray`, `ImmutableSparseNDimArray` , removing it to `ImmutableNDimArray` will avoid repetitive code.